src/util/merkleTree.ts

```diff
@@ -3,126 +3,135 @@
  */
 
 import assert from "assert";
-import {bytes32} from "../types";
-import {hash} from "../util/crypto";
-import {intDiv} from "../util/math";
+import {bytes, bytes32, number64} from "../types";
+import {hash} from "./crypto";
+import {intDiv} from "./math";
+import {serialize} from "@chainsafe/ssz";
+import {MerkleTree} from "../types";
+
+export interface IProgressiveMerkleTree {
 
-interface IProgressiveMerkleTree {
-  /**
-   * The number of items in the tree
-   */
-  count(): number;
   depth(): number;
+
   /**
-   * Add a new item into the tree
-   *
-   * Return a proof to verify the item is in the tree
-   * @returns proof
+   * push new item into the tree
    */
-  push(item: bytes32): bytes32[];
+  push(item: bytes32): void;
+
+  add(index: number64, item: bytes32): void;
+
+  getProof(index: number64): bytes32[];
+
   /**
    * The merkle root of the tree
    */
   root(): bytes32;
+
+  serialize(): Buffer;
 }
 
 export class ProgressiveMerkleTree implements IProgressiveMerkleTree {
```

Comment:
Can we keep the old implementation around? It might still be useful eventually.

Reply:
You kinda have all old implementation features here.

Reply:
I think this uses a lot more memory (because it has to keep the deposit roots around).
I was thinking it would be useful to keep the lightweight version around for times when we don't need to rerun proofs for old deposits against new roots.

Reply:
I saved old implementation